### DISCUSSION TITLE
Websocket client on STM32 board <--> websocket server on raspberry pi

### DISCUSSION THREAD

**Discussion by Manfred-O** [cesanta/mongoose#2992] [Ask questions, get answers] [answered]
Hi, i have following setup.
I'm using both tutorials for websockets server&client. The client side build fine on a stm32f439. Got my IP etc.. 
The server is setup on a raspberry pi which is connected to my local network via vpn. I started the server and got "WS upgrade expected" when using chrome also the filesystem is showing up.
But my stm32 board run into timeout. Any advice.

**Comment by scaprile** [comment, marked as answer]
Please take the time to learn the technology you are using.
Please read the tutorials again, understand them, do the examples as suggested; we never access a WebSocket server with an HTTP client, thing you are doing.
A WebSocket server "upgrades" an HTTP connection, that is what you are seeing because you are not using a WebSocket client (thing we do in our tutorials; in fact, there is a JS client in the server example https://github.com/cesanta/mongoose/blob/master/tutorials/websocket/websocket-server/test.html and there is an indication in the tutorial to use it)
![yab](https://github.com/user-attachments/assets/a4f59176-f091-4071-84ba-5d77415a1651)

https://github.com/cesanta/mongoose/blob/7b296fabb0ea04c68d7cc3cbbbaca64526b767d5/tutorials/websocket/websocket-server/README.md?plain=1#L1
https://mongoose.ws/documentation/tutorials/websocket/websocket-server/


Whatever else you have on your network or your VPN is beyond our analysis, you should know your network.
If you want further help, do proper analysis: sniff your network and ask precise questions.
